File: cpp/contains-duplicate-iii.cpp
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
    bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {
		typedef long long ll;
		int N = nums.size();
		if (k < 1) return false;

		set<ll> window;
		int i = 0;
		for (int j = 0; j < N; ++j) {
			if (j - i > k) {
				window.erase(nums[i]);
				i++;
			}

			// nums[j] - t <= nums[i] <= nums[j] + t
			// window is increase sort
			auto it = window.lower_bound((ll)nums[j] - t);
			if (it != window.end() && abs((ll)nums[j] - *it) <= t) return true;

			window.insert(nums[j]);
		}

		return false;
    }

    bool containsNearbyAlmostDuplicate2(vector<int>& nums, int k, int t) {
		int N = nums.size();
		if (k <= 0) return false;
		for (int i = 1; i <= k; ++i) {
			for (int j = 0; j+i < N; ++j) {
				if (abs(nums[j+i] - nums[j]) <= t) return true;
			}
		}
		return false;
    }
};

int main(int argc, char *argv[]) {
	Solution s;
	{
		vector<int> v_nums = {4,2};
		cout << s.containsNearbyAlmostDuplicate(v_nums, 2, 3) <<endl;
	}
	{
		vector<int> v_nums = {4,2};
		cout << s.containsNearbyAlmostDuplicate(v_nums, 2, 1) <<endl;
	}
	{
		vector<int> v_nums = {-1, -1};
		cout << s.containsNearbyAlmostDuplicate(v_nums, 1, 0) <<endl;
	}
	return 0;
}

File: js/contains-duplicate-iii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @param {number} t
 * @return {boolean}
 */

var containsNearbyAlmostDuplicate = function(nums, k, t) {
  var N = nums.length;
  if (N <= 1 || k < 1) return false;

  var pairs = [];

  for (var i = 0; i < N; i++) {
	pairs.push([nums[i], i]);
  }

  pairs.sort(function(a, b) { return a[0] - b[0]; });

  for (i = 0; i < N; i++) {
	var j = i+1;
	while (j < N) {
	  if (Math.abs(pairs[i][0] - pairs[j][0]) > t) break;
	  if (Math.abs(pairs[i][1] - pairs[j][1]) <= k) return true;
	  j++;
	}
  }

  return false;
};

{
  var nums = [4, 2];
  console.log(containsNearbyAlmostDuplicate(nums, 2, 3));
}

{
  var nums = [-1, -1];
  console.log(containsNearbyAlmostDuplicate(nums, 1, -1));
}
